ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,670 @@
+pub const SIGNER_VANITY_LENGTH: usize = 32;
+/// Fixed number of extra-data suffix bytes reserved for signer signature
+pub const SIGNER_SIG_LENGTH: usize = 65;
+/// Nonce value for DROP vote
+pub const NONCE_DROP_VOTE: [u8; 8] = [0x00; 8];
+/// Nonce value for AUTH vote
+pub const NONCE_AUTH_VOTE: [u8; 8] = [0xff; 8];
+/// Difficulty for INTURN block
+pub const DIFF_INTURN: U256 = U256([2, 0, 0, 0]);
+/// Difficulty for NOTURN block
+pub const DIFF_NOTURN: U256 = U256([1, 0, 0, 0]);
+/// Default empty author field value
+pub const NULL_AUTHOR: Address = H160([0x00; 20]);
+/// Default empty nonce value
+pub const NULL_NONCE: [u8; 8] = NONCE_DROP_VOTE;
+/// Default value for mixhash
+pub const NULL_MIXHASH: [u8; 32] = [0x00; 32];
+/// Default value for uncles hash
+pub const NULL_UNCLES_HASH: H256 = KECCAK_EMPTY_LIST_RLP;
+/// Default noturn block wiggle factor defined in spec.
+pub const SIGNING_DELAY_NOTURN_MS: u64 = 500;
+
+#[derive(PartialEq, Clone, Debug, Copy)]
+pub enum VoteType {
+	Add,
+	Remove,
+}
+
+// Caches
+/// How many CliqueBlockState to cache in the memory.
+pub const STATE_CACHE_NUM: usize = 128;
+
+/// Clique Engine implementation
+/// block_state_by_hash -> block state indexed by header hash.
+pub struct Clique {
+	epoch_length: u64,
+	period: u64,
+	machine: EthereumMachine,
+	client: RwLock<Option<Weak<EngineClient>>>,
+	block_state_by_hash: RwLock<LruCache<H256, CliqueBlockState>>,
+	proposals: RwLock<HashMap<Address, VoteType>>,
+	signer: RwLock<Option<Box<EngineSigner>>>,
+	step_service: Option<Arc<StepService>>,
+}
+
+impl Clique {
+	/// Initialize Clique engine from empty state.
+	pub fn new(our_params: CliqueParams, machine: EthereumMachine) -> Result<Arc<Self>, Error> {
+		let mut engine = Clique {
+			epoch_length: our_params.epoch,
+			period: our_params.period,
+			client: RwLock::new(Option::default()),
+			block_state_by_hash: RwLock::new(LruCache::new(STATE_CACHE_NUM)),
+			proposals: RwLock::new(Default::default()),
+			signer: RwLock::new(Default::default()),
+			machine,
+			step_service: None,
+		};
+
+		let res =  Arc::new(engine);
+
+		if our_params.period > 0 {
+			engine.step_service = Some(StepService::start(Arc::downgrade(&res) as Weak<Engine<_>>));
+		}
+
+		Ok(res)
+	}
+
+	fn sign_header(&self, header: &Header) -> Result<(Signature, H256), Error> {
+		match self.signer.read().as_ref() {
+			None => {
+				return Err(Box::new("sign_header: No signer available.").into());
+			}
+			Some(signer) => {
+				let digest = header.hash();
+				match signer.sign(digest) {
+					Ok(sig) => {
+						return Ok((sig, digest));
+					}
+					Err(e) => {
+						return Err(Box::new(format!("sign_header: failed to sign header, error: {}", e)).into());
+					}
+				}
+			}
+		}
+	}
+
+	/// Construct an new state from given checkpoint header.
+	fn new_checkpoint_state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		assert_eq!(header.number() % self.epoch_length, 0);
```

Comment:
should this be enabled in `release builds` i.e, will panic?

I suggest making this `debug_assert_eq` or returning `Err` because it a logic bug?

I can see that it is called from three places and I can verify that one is checked AFAIK.